src/collapse/common.rs

```diff
@@ -0,0 +1,794 @@
+/// for free as long as you adhere to the requirements described in the
+/// comments below.
+pub trait CollapsePrivate: Clone + Send + Sized + Sized {
+    // *********************************************************** //
+    // ********************* REQUIRED METHODS ******************** //
+    // *********************************************************** //
+
+    /// Some formats, such as `dtrace`, contain a header or other non-stack
+    /// information at the beginning of their input files. If header information
+    /// is present, this method **must** consume it (i.e. advance the provided
+    /// reader past it).
+    ///
+    /// This method also provides an opportunity to do processing of actual
+    /// stack data on the main thread before worker threads are spun up. For
+    /// an example of why this might be necessary, see `perf`, which can require
+    /// reading the first stack in order to know how to process the rest.
+    ///
+    /// If the format you are working with does not contain header information
+    /// or does not need any special, up-front processing, just have this method
+    /// return `Ok(())` immediately.
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The primary method.
+    ///
+    /// It receives a reader whose header has already been consumed (see above), as
+    /// well as a mutable reference to an `Occurences` instance (just a hashamp that
+    /// works across multiple threads). Implementators should parse the stack data
+    /// contained in the reader and write output to the provided `Occurrences` map.
+    ///
+    /// So that this method may be called multiple times, all internal
+    /// state contained in `self` (e.g. any stack buffers or the like) **must** be
+    /// reset before this method returns. Also, this method may **not** use threads.
+    fn collapse_single_threaded<R>(
+        &mut self,
+        reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The following method is used to determine how to chunk up the input data in order to
+    /// send it off to worker threads, which **must** receive full stacks (as opposed to partial
+    /// stacks). Since what comprises a "stack" will vary from format to format, implementors for
+    /// a specific format need to implement this method, which encodes how we know when we've
+    /// reached the end of a stack.
+    ///
+    /// This method **must** return `true` if the provided line represents the end of a stack;
+    /// `false` otherwise.
+    ///
+    /// If your format requires more information than merely a line of the input data in order
+    /// to determine whether or not you are at the end of a stack, you can access information
+    /// stored on the `self` instance, which is also available to you in this method.
```

Comment:
Unless I'm missing something, I think that would require `&mut self`, otherwise how does it maintain state across multiple consecutive calls? I don't _think_ any other method is called for each line?

Reply:
Totally correct. Oversight on my part.